src/libgrate/grate-3d.h

```diff
@@ -0,0 +1,115 @@
+#ifndef GRATE_LIBGRATE_3D_H
+#define GRATE_LIBGRATE_3D_H 1
+
+#include <stdint.h>
+
+struct grate_bo;
+struct cgc_shader;
+
+struct grate_attribute {
+	int position;
+	const char *name;
+};
+
+struct grate_uniform {
+	int position;
+	const char *name;
+};
+
+struct grate_shader {
+	struct cgc_shader *cgc;
+	unsigned num_words;
+	uint32_t *words;
+
+	union {
+		struct {
+			unsigned alu_buf_size;
+			unsigned pseq_inst_nb;
+			unsigned pseq_to_dw_nb;
+		};
+
+		struct {
+			unsigned used_tram_rows_nb;
+			unsigned linker_inst_nb;
+		};
+	};
+};
+
+struct grate_program {
+	struct grate_shader *vs;
+	struct grate_shader *fs;
+	struct grate_shader *linker;
+
+	struct grate_attribute *attributes;
+	unsigned num_attributes;
+	uint32_t attributes_use_mask;
+
+	struct grate_uniform *vs_uniforms;
+	unsigned num_vs_uniforms;
+
+	struct grate_uniform *fs_uniforms;
+	unsigned num_fs_uniforms;
+
+	uint32_t vs_constants[256 * 4];
+	uint32_t fs_constants[32];
+};
+
+struct grate_render_target {
+	struct grate_bo *bo;
+	unsigned pixel_format;
+	unsigned pitch;
+	bool dither_enable;
```

Comment:
Yeah, I changed it in in the follow-up commit https://github.com/grate-driver/grate/pull/13/commits/c05db08346a2d9e194b6324a13be6883de57f1dd